<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.4.0 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10222-J3VBQ4K5JK33FLXPO3XIQIBUQG3YHYHUZKUN4V5KO7OAHDN2V2DQ
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

Running "checkFormat:ios" (checkFormat) task

Running "checkFormat:android" (checkFormat) task
Fatal error: Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/io/TiInputStreamWrapper.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='321' length='3'>&#10;&#10;</replacement>
<replacement offset='761' length='1'>&#10;</replacement>
<replacement offset='987' length='4'>&#10;&#10;	</replacement>
<replacement offset='1175' length='4'>&#10;&#10;	</replacement>
<replacement offset='1544' length='1'>&#10;	</replacement>
<replacement offset='1998' length='1'>&#10;	</replacement>
<replacement offset='2224' length='1'>&#10;	</replacement>
<replacement offset='2866' length='1'>&#10;	</replacement>
<replacement offset='3166' length='1'>&#10;	</replacement>
<replacement offset='3650' length='1'>&#10;	</replacement>
<replacement offset='5530' length='1'>&#10;	</replacement>
<replacement offset='5886' length='1'>&#10;	</replacement>
<replacement offset='6514' length='1'>&#10;	</replacement>
</replacements>

Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/util/TiDownloadManager.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='4068' length='1'>&#10;			</replacement>
<replacement offset='4158' length='1'>&#10;				</replacement>
<replacement offset='4224' length='1'>&#10;				</replacement>
<replacement offset='4483' length='1'>&#10;				</replacement>
<replacement offset='5055' length='5'> </replacement>
<replacement offset='7545' length='5'> </replacement>
<replacement offset='8058' length='1'>&#10;				</replacement>
<replacement offset='8113' length='6'> </replacement>
</replacements>

Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/util/TiImageHelper.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='5029' length='1'>&#10;							 </replacement>
<replacement offset='5031' length='5'> </replacement>
<replacement offset='6523' length='1'>&#10;					</replacement>
<replacement offset='6584' length='18'> </replacement>
<replacement offset='6763' length='2'> </replacement>
<replacement offset='7058' length='2'> </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

--- a/android/titanium/src/java/org/appcelerator/titanium/io/TiInputStreamWrapper.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/io/TiInputStreamWrapper.java
@@ -10,7 +10,6 @@ package org.appcelerator.titanium.io;
 import java.io.InputStream;
 import java.io.IOException;
 
-
 /**
  * Wraps an input stream, passing the calls through from this wrapper to the wrapped stream.
  * <p>
@@ -18,13 +17,13 @@ import java.io.IOException;
  * allowing the creator of the wrapper to perform additional cleanup. For example, the listener
  * could close an "HttpURLConnection" or "ZipFile" after the stream has been closed.
  */
-public class TiInputStreamWrapper extends InputStream {
+public class TiInputStreamWrapper extends InputStream
+{
 	/** The input stream that is being wrapped. Can be null. */
 	private InputStream inputStream;
 
 	/** The listener to be invoked after the input stream has been closed. Can be null. */
 	private ClosedListener closedListener;
-
 
 	/** Listener to be invoked after the input stream has been closed. */
 	public interface ClosedListener {
@@ -32,13 +31,13 @@ public class TiInputStreamWrapper extends InputStream {
 		void onClosed();
 	}
 
-
 	/**
 	 * Creates a new input stream wrapping the given input stream.
 	 * @param inputStream The stream to be wrapped. Can be null, in which case, this wrapper's methods will no-op.
 	 * @param closedListener Listener to be invoked after the stream has been closed. Can be null.
 	 */
-	public TiInputStreamWrapper(InputStream inputStream, ClosedListener closedListener) {
+	public TiInputStreamWrapper(InputStream inputStream, ClosedListener closedListener)
+	{
 		super();
 
 		this.inputStream = inputStream;
@@ -54,7 +53,8 @@ public class TiInputStreamWrapper extends InputStream {
 	 * or if the wrapper was given a null stream reference.
 	 */
 	@Override
-	public int available() throws IOException {
+	public int available() throws IOException
+	{
 		if (this.inputStream != null) {
 			return this.inputStream.available();
 		}
@@ -63,7 +63,8 @@ public class TiInputStreamWrapper extends InputStream {
 
 	/** Closes the input stream and releases any resources associated with it. */
 	@Override
-	public void close() throws IOException {
+	public void close() throws IOException
+	{
 		// Close the input stream.
 		if (this.inputStream != null) {
 			this.inputStream.close();
@@ -83,7 +84,8 @@ public class TiInputStreamWrapper extends InputStream {
 	 * @param The maximum number of bytes that can be read before the mark position becomes invalid.
 	 */
 	@Override
-	public void mark(int readLimit) {
+	public void mark(int readLimit)
+	{
 		if (this.inputStream != null) {
 			this.inputStream.mark(readLimit);
 		}
@@ -94,7 +96,8 @@ public class TiInputStreamWrapper extends InputStream {
 	 * @return Returns true if the mark() and reset() methods are supported. Returns false if not.
 	 */
 	@Override
-	public boolean markSupported() {
+	public boolean markSupported()
+	{
 		if (this.inputStream != null) {
 			return this.inputStream.markSupported();
 		}
@@ -110,7 +113,8 @@ public class TiInputStreamWrapper extends InputStream {
 	 * Returns -1 if the end of the stream was reached or if the wrapper was given a null stream reference.
 	 */
 	@Override
-	public int read() throws IOException {
+	public int read() throws IOException
+	{
 		if (this.inputStream != null) {
 			return this.inputStream.read();
 		}
@@ -155,7 +159,8 @@ public class TiInputStreamWrapper extends InputStream {
 	 * Returns -1 if there is no more data to be read from the stream.
 	 */
 	@Override
-	public int read(byte[] byteBuffer) throws IOException, NullPointerException {
+	public int read(byte[] byteBuffer) throws IOException, NullPointerException
+	{
 		if (this.inputStream != null) {
 			return this.inputStream.read(byteBuffer);
 		}
@@ -168,7 +173,8 @@ public class TiInputStreamWrapper extends InputStream {
 	 * This method does nothing if the stream's markSupported() method returns false.
 	 */
 	@Override
-	public void reset() throws IOException {
+	public void reset() throws IOException
+	{
 		if (this.inputStream != null) {
 			this.inputStream.reset();
 		}
@@ -185,7 +191,8 @@ public class TiInputStreamWrapper extends InputStream {
 	 * Will also return zero if this wrapper was given a null stream reference.
 	 */
 	@Override
-	public long skip(long byteCount) throws IOException {
+	public long skip(long byteCount) throws IOException
+	{
 		if (this.inputStream != null) {
 			return this.inputStream.skip(byteCount);
 		}

--- a/android/titanium/src/java/org/appcelerator/titanium/util/TiDownloadManager.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/util/TiDownloadManager.java
@@ -88,7 +88,7 @@ public class TiDownloadManager implements Handler.Callback
 	 * <p>
 	 * Returns null if failed to download content or if given an invalid argument.
 	 */
-	public InputStream blockingDownload(URI uri)
+	public InputStream blockingDownload(final URI uri)
 	{
 		// Validate.
 		if (uri == null) {
@@ -112,44 +112,24 @@ public class TiDownloadManager implements Handler.Callback
 		// If running on main UI thread, then block this thread while doing download on another thread.
 		// Note: Using "HttpUrlConnection" on UI thread will cause a "NetworkOnMainThreadException" to be thrown.
 		if (TiApplication.isUIThread()) {
-			// Anonymous class used to call this blockingDownload() method on another thread.
+			// Perform the blocking download on another thread.
 			// Downloaded content will be made available via Titanium's "TiResponseCache".
-			class DownloadOperation implements Runnable
-			{
-				private boolean isDone;
-				private URI uri;
-
-				public DownloadOperation(URI uri)
-				{
-					this.uri = uri;
-				}
-
-				@Override
-				public void run()
-				{
-					try (InputStream stream = blockingDownload(this.uri)) {
-						if (stream != null) {
-							KrollStreamHelper.pump(stream, null);
-						}
-					} catch (Exception ex) {
-					} finally {
-						this.isDone = true;
-					}
-				}
-
-				public boolean isDone()
-				{
-					return this.isDone;
-				}
-			};
-
-			// Perform the blocking download.
 			try {
-				DownloadOperation downloadOperation = new DownloadOperation(uri);
-				(new Thread(downloadOperation)).start();
-				while (downloadOperation.isDone() == false) {
-					Thread.sleep(10);
-				}
+				Thread thread = new Thread(new Runnable() {
+					@Override
+					public void run()
+					{
+						try (InputStream stream = blockingDownload(uri)) {
+							if (stream != null) {
+								KrollStreamHelper.pump(stream, null);
+							}
+						} catch (Exception ex) {
+							Log.e(TAG, "Exception downloading from: " + uri.toString(), ex);
+						}
+					}
+				});
+				thread.start();
+				thread.join();
 			} catch (Exception ex) {
 			}
 
@@ -170,7 +150,7 @@ public class TiDownloadManager implements Handler.Callback
 		try {
 			url = new URL(uri.toString());
 		} catch (Exception ex) {
-			ex.printStackTrace();
+			Log.e(TAG, "Failed to parse URL: " + uri.toString(), ex);
 			return null;
 		}
 
@@ -191,12 +171,12 @@ public class TiDownloadManager implements Handler.Callback
 					httpConnection.connect();
 					int responseCode = httpConnection.getResponseCode();
 					if (responseCode == 200) {
-						// A valid respoonse has been received. Fetch its download stream.
+						// A valid response has been received. Fetch its download stream.
 						inputStream = httpConnection.getInputStream();
 					} else if ((responseCode >= 300) && (responseCode < 400)) {
 						// A redirect response has been received.
-						// First, make sure we haven't been redirected too many times.
-						// Note: This is a rudimentary check to avoid recursive redirects.
+						// We have to handle redirects between HTTP/HTTPS protocols ourselves.
+						// First, make sure we haven't been redirected too many times. (Avoids recursive redirects.)
 						redirectCount++;
 						if (redirectCount > 10) {
 							break;
@@ -212,6 +192,9 @@ public class TiDownloadManager implements Handler.Callback
 						// Close the connection and attempt to connect to the redirect URL.
 						httpConnection.disconnect();
 						continue;
+					} else {
+						// Server did not provide the expected response, such as a 404.
+						Log.e(TAG, "Received unexpected response code " + responseCode + " from: " + uri);
 					}
 				} else {
 					// Connect to the endpoint and acquire the downloaded content.
@@ -364,7 +347,7 @@ public class TiDownloadManager implements Handler.Callback
 
 				// fire a download fail event if we are unable to download
 				sendMessage(uri, MSG_FIRE_DOWNLOAD_FAILED);
-				Log.e(TAG, "Exception downloading " + uri, e);
+				Log.e(TAG, "Exception downloading from: " + uri, e);
 			}
 		}
 	}

--- a/android/titanium/src/java/org/appcelerator/titanium/util/TiDownloadManager.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/util/TiDownloadManager.java
@@ -114,16 +114,19 @@ public class TiDownloadManager implements Handler.Callback
 		if (TiApplication.isUIThread()) {
 			// Anonymous class used to call this blockingDownload() method on another thread.
 			// Downloaded content will be made available via Titanium's "TiResponseCache".
-			class DownloadOperation implements Runnable {
+			class DownloadOperation implements Runnable
+			{
 				private boolean isDone;
 				private URI uri;
 
-				public DownloadOperation(URI uri) {
+				public DownloadOperation(URI uri)
+				{
 					this.uri = uri;
 				}
 
 				@Override
-				public void run() {
+				public void run()
+				{
 					try (InputStream stream = blockingDownload(this.uri)) {
 						if (stream != null) {
 							KrollStreamHelper.pump(stream, null);
@@ -134,7 +137,8 @@ public class TiDownloadManager implements Handler.Callback
 					}
 				}
 
-				public boolean isDone() {
+				public boolean isDone()
+				{
 					return this.isDone;
 				}
 			};
@@ -154,8 +158,7 @@ public class TiDownloadManager implements Handler.Callback
 			if (cachedUri != null) {
 				try {
 					inputStream = TiResponseCache.openCachedStream(cachedUri);
-				}
-				catch (Exception ex) {
+				} catch (Exception ex) {
 				}
 			}
 			return inputStream;
@@ -226,8 +229,7 @@ public class TiDownloadManager implements Handler.Callback
 			if ((inputStream == null) && (connection instanceof HttpURLConnection)) {
 				try {
 					((HttpURLConnection) connection).disconnect();
-				}
-				catch (Exception ex) {
+				} catch (Exception ex) {
 				}
 			}
 		}
@@ -238,11 +240,11 @@ public class TiDownloadManager implements Handler.Callback
 			final HttpURLConnection httpConnection = (HttpURLConnection) connection;
 			inputStream = new TiInputStreamWrapper(inputStream, new TiInputStreamWrapper.ClosedListener() {
 				@Override
-				public void onClosed() {
+				public void onClosed()
+				{
 					try {
 						httpConnection.disconnect();
-					}
-					catch (Exception ex) {
+					} catch (Exception ex) {
 					}
 				}
 			});

--- a/android/titanium/src/java/org/appcelerator/titanium/util/TiImageHelper.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/util/TiImageHelper.java
@@ -150,8 +150,8 @@ public class TiImageHelper
 	{
 		// Validate argument.
 		if ((path == null) || path.isEmpty()) {
-			String message = "Path of image file could not determined. " +
-				"Could not create an exifInterface from an invalid path.";
+			String message = "Path of image file could not determined. "
+							 + "Could not create an exifInterface from an invalid path.";
 			Log.e(TAG, message);
 			return 0;
 		}
@@ -193,11 +193,11 @@ public class TiImageHelper
 		try {
 			if (stream != null) {
 				ExifInterface exifInterface = new ExifInterface(stream);
-				int exifOrientation = exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION,
-																	ExifInterface.ORIENTATION_NORMAL);
+				int exifOrientation =
+					exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL);
 				switch (exifOrientation) {
 					case ExifInterface.ORIENTATION_ROTATE_270:
-					case ExifInterface.ORIENTATION_TRANSVERSE:		// Rotated and mirrored.
+					case ExifInterface.ORIENTATION_TRANSVERSE: // Rotated and mirrored.
 						orientation = 270;
 						break;
 					case ExifInterface.ORIENTATION_ROTATE_180:
@@ -205,7 +205,7 @@ public class TiImageHelper
 						orientation = 180;
 						break;
 					case ExifInterface.ORIENTATION_ROTATE_90:
-					case ExifInterface.ORIENTATION_TRANSPOSE:		// Rotated and mirrored.
+					case ExifInterface.ORIENTATION_TRANSPOSE: // Rotated and mirrored.
 						orientation = 90;
 						break;
 				}

